docs/overriding_agg_fn.rst

```diff
@@ -32,12 +32,125 @@ Choose from the following predefined aggregation functions:
+    the user must specify parameters for the aggregation optimizer
+    (``NumPyAdagrad``, ``NumPyAdam`` or ``NumPyYogi``) that will aggregate
+    the global model. Theese parameters parameters are passed via **keywords**.
+
+    Also, user must pass one of the arguments: ``params``
+    - model parameters (a dictionary with named model parameters
+    in the form of numpy arrays), or pass ``model_interface``
+    - an instance of the `ModelInterface <https://github.com/intel/openfl/blob/develop/openfl/interface/interactive_api/experiment.py>`_ class.
+    If user pass both ``params`` and ``model_interface``,
+    then the optimizer parameters are initialized via
+    ``params``, ignoring ``model_interface`` argument.
+
+    See the `AdagradAdaptiveAggregation
+    <https://github.com/intel/openfl/blob/develop/openfl/component/aggregation_functions/adagrad_adaptive_aggregation.py>`_
+    definitions for details.
+
+    `Adaptive federated optimization <https://arxiv.org/pdf/2003.00295.pdf>`_ original paper.
+
+``AdagradAdaptiveAggregation`` usage example:
+
+.. code-block:: python
+
+    from openfl.interface.interactive_api.experiment import TaskInterface, ModelInterface
+    from openfl.component.aggregation_functions import AdagradAdaptiveAggregation
+
+    TI = TaskInterface()
+    MI = ModelInterface(model=model,
+                        optimizer=optimizer,
+                        framework_plugin=framework_adapter)
+    ...
+
+    # Creating aggregation function
+    agg_fn = AdagradAdaptiveAggregation(model_interface=MI,
+                                        learning_rate=0.4)
+
+    # Define training task
+    @TI.register_fl_task(model='model', data_loader='train_loader', \
+                            device='device', optimizer='optimizer')
+    @TI.set_aggregation_function(agg_fn)
+    def train(...):
+    ...
+
+You can define your own numpy based optimizer,
+which will be used for global model aggreagation:
+
+.. code-block:: python
+
+    from openfl.utilities.optimizers.numpy.base_optimizer import Optimizer
+
+    class MyOpt(Optimizer):
+        """My optimizer implementation."""
+
+        def __init__(
+            self,
+            *,
+            params: Optional[Dict[str, np.ndarray]] = None,
+            model_interface=None,
+            learning_rate: float = 0.001,
+            param1: Any = None,
+            param2: Any = None
+        ) -> None:
+            """Initialize.
+
+            Args:
+                params: Parameters to be stored for optimization.
+                model_interface: Model interface instance to provide parameters.
+                learning_rate: Tuning parameter that determines
+                    the step size at each iteration.
+                param1: My own defined parameter.
+                param2: My own defined parameter.
+            """
+            super().__init__()
+            pass # Your code here!
+
+        def step(self, gradients: Dict[str, np.ndarray]) -> None:
```

Comment:
could you provide a scenario when `gradients` is a string?

Reply:
@igor-davidyuk According to mypy [docs](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html):

```python3
# For mappings, we need the types of both keys and values
x: dict[str, float] = {"field": 2.0}  # Python 3.9+
x: Dict[str, float] = {"field": 2.0}
```